File: app/ios/OffTheCloud/OffTheCloud/WSClient.swift
// SPDX-License-Identifier: AGPL-3.0-or-later

//
//  WSClient.swift
//  OffTheCloud
//
//  Created by Alonso Vidales on 8/9/25.
//

import Foundation

/// An `actor` (not a plain class) because URLSession delivers `receive`/`send`
/// completions on its own background queue, not on whatever thread called
/// into this client. With a plain class those callbacks mutated `waiters`/
/// `nextId` concurrently with calls made from OTCConnection, a real data
/// race that showed up in practice as a heap-corruption crash on a real
/// device once traffic got busy (large file listings). The actor serializes
/// every access to this state, callbacks included.
actor WSClient {
    private var task: URLSessionWebSocketTask?
    private let session: URLSession
    private(set) var connected = false
    private var nextId: Int32 = 1
    private var waiters = [Int32: (Result<Msg_RespEnvelope, Error>) -> Void]()

    /// Fired once, from the receive loop, when the socket breaks (read error
    /// or clean close). The owner (OTCConnection) uses this to mark itself
    /// unauthenticated and schedule a reconnect; WSClient itself does not
    /// retry on its own.
    var onDisconnect: (() -> Void)?

    init() {
        let cfg = URLSessionConfiguration.default
        cfg.waitsForConnectivity = true
        session = URLSession(configuration: cfg)
    }

    func setOnDisconnect(_ cb: @escaping () -> Void) {
        onDisconnect = cb
    }

    func connect(url: URL) async throws {
        print("Trynig to connect to: \(url)")
        if let task, task.state == .running { return }
        let t = session.webSocketTask(with: url)
        // The default is 1 MiB, which a file listing for a few thousand
        // photos blows straight past — every receive then fails with
        // "Message too long" and the connection never gets anywhere. Match
        // the macOS client's generous cap.
        t.maximumMessageSize = 1000 * 1024 * 1024
        task = t
        t.resume()
        connected = true
        print("Connected!!!")
        listen()
    }

    func close() {
        task?.cancel()
        connected = false
    }

    private func listen() {
        task?.receive { [weak self] result in
            guard let self else { return }
            Task { await self.handleReceive(result) }
        }
    }

    private func handleReceive(_ result: Result<URLSessionWebSocketTask.Message, Error>) {
        switch result {
        case .failure(let err):
            print("WS receive error:", err)
            failAndClose(err)
        case .success(let message):
            switch message {
            case .data(let data):
                do {
                    let env = try Msg_RespEnvelope(serializedData: data)
                    if let cb = waiters[env.id] {
                        waiters.removeValue(forKey: env.id)
                        cb(.success(env))
                    }
                    // If you expect server push, also post a Notification here.
                } catch {
                    print("Decode error:", error)
                }
            default: break
            }
            // keep listening only while the socket is still healthy
            listen()
        }
    }

    /// Marks the connection dead, fails every outstanding request, and
    /// notifies the owner so it can reconnect. Does NOT call listen() again
    /// — a fresh connect() starts a new receive loop.
    private func failAndClose(_ error: Error) {
        connected = false
        let pending = waiters
        waiters.removeAll()
        for (_, cb) in pending { cb(.failure(error)) }
        task?.cancel(with: .abnormalClosure, reason: nil)
        onDisconnect?()
    }

    func request(build: (inout Msg_ReqEnvelope) -> Void) async throws -> Msg_RespEnvelope {
        guard connected else {
            throw NSError(domain: "ws", code: -1, userInfo: [NSLocalizedDescriptionKey: "Not connected"])
        }
        var env = Msg_ReqEnvelope()
        let id = nextId; nextId += 1
        env.id = id
        build(&env)
        // Re-assert the id after build() rather than trusting it survived:
        // this is issue #77's actual "stuck loading" bug. Most call sites'
        // closures do `var req = ReqEnvelope(); ...; e = req`, replacing
        // the whole envelope - id included - with a fresh one whose id
        // defaults to 0, instead of mutating the passed-in `e` in place.
        // Every request built that way was silently going out on the wire
        // as id=0, so any two of them in flight at once collided on the
        // same waiters[0] slot: the second registration overwrote the
        // first's, and when the first request's actual response
        // eventually arrived, waiters[0] pointed at someone else's
        // callback, leaving the first one's continuation waiting forever.
        // Confirmed live via the Swift runtime's own "leaked its
        // continuation" diagnostic and by literally logging "request
        // id=0" for nearly every request. Restoring the real id here
        // fixes every call site at once, rather than relying on dozens of
        // them to each preserve it correctly on their own.
        env.id = id

        //print("Sending request: \(env)")

        let data = try env.serializedData()
        return try await withCheckedThrowingContinuation { cont in
            self.waiters[id] = { result in cont.resume(with: result) }
            self.task?.send(.data(data)) { error in
                guard let error else { return }
                Task { await self.failWaiter(id: id, error: error) }
            }
        }
    }

    private func failWaiter(id: Int32, error: Error) {
        if let cb = waiters.removeValue(forKey: id) {
            cb(.failure(error))
        }
    }
}

File: app/android/app/src/main/java/cloud/offthe/otc/net/WSClient.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package cloud.offthe.otc.net

import cloud.offthe.otc.proto.ReqEnvelope
import cloud.offthe.otc.proto.RespEnvelope
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

// Port of WSClient.swift: one WebSocket, requests correlated with
// responses by envelope id. OkHttp delivers listener callbacks on its own
// threads, so every access to the waiters map goes through the mutex -
// the same reason the Swift version is an actor.
class WSClient {
    private val client = OkHttpClient.Builder()
        .pingInterval(30, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS) // long-lived socket
        .build()

    private var socket: WebSocket? = null
    @Volatile var connected = false
        private set
    private var nextId = 1
    private val waiters = HashMap<Int, (Result<RespEnvelope>) -> Unit>()
    private val lock = Mutex()

    /** Fired once, when the socket breaks. The owner reconnects; this class never retries. */
    var onDisconnect: (() -> Unit)? = null

    suspend fun connect(url: String) {
        if (connected) return
        val opened = CompletableDeferred<Unit>()
        val req = Request.Builder().url(url).build()
        socket = client.newWebSocket(req, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                connected = true
                opened.complete(Unit)
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                val env = try { RespEnvelope.parseFrom(bytes.toByteArray()) } catch (e: Exception) { return }
                val cb = synchronized(waiters) { waiters.remove(env.id) }
                cb?.invoke(Result.success(env))
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                val err = if (response != null && !connected) {
                    IOException("bad response from the server (${response.code})", t)
                } else t
                if (!opened.isCompleted) opened.completeExceptionally(err)
                failAndClose(err)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(code, reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                failAndClose(IOException("connection closed ($code)"))
            }
        })
        opened.await()
    }

    fun close() {
        connected = false
        socket?.cancel()
        socket = null
    }

    private fun failAndClose(error: Throwable) {
        if (!connected && socket == null) return
        connected = false
        val pending = synchronized(waiters) { val p = waiters.values.toList(); waiters.clear(); p }
        pending.forEach { it(Result.failure(error)) }
        socket?.cancel()
        socket = null
        onDisconnect?.invoke()
    }

    suspend fun request(build: (ReqEnvelope.Builder) -> Unit): RespEnvelope {
        if (!connected) throw IOException("Not connected")
        val id = lock.withLock { nextId++ }
        val b = ReqEnvelope.newBuilder()
        build(b)
        // Re-assert the id after build(), as the Swift client does (issue
        // #77): a call site that replaces the whole envelope would send
        // id 0 and collide with every other in-flight request.
        b.id = id
        val bytes = b.build().toByteArray()
        return suspendCancellableCoroutine { cont ->
            synchronized(waiters) {
                waiters[id] = { r -> r.fold({ cont.resume(it) }, { cont.resumeWithException(it) }) }
            }
            val ok = socket?.send(bytes.toByteString()) ?: false
            if (!ok) {
                val cb = synchronized(waiters) { waiters.remove(id) }
                cb?.invoke(Result.failure(IOException("send failed")))
            }
            cont.invokeOnCancellation { synchronized(waiters) { waiters.remove(id) } }
        }
    }
}
